### PR TITLE
feat(profiling): keep trace resource updated in profiler events

### PR DESCRIPTION
**Pull request by jd**
The current way of storing traces and spans information in the profiler events
do not guarantee that the updates done to some of the trace/span attributes is
being kept up-to-date.

The typical example is the root span resource which might be updated just
before the trace is finished. For the profiler, this is too late as the trace
resource is already copied into the sampling events it generated.

This change leverages the span finish callback mechanism to update the profiler
events when the span is finished. This synchronizes back the trace information
from the trace to the event linked to it.

There's still a small chance that the profiling event that is updated is
already flushed out. In that case, the update will be useless. This can happen
if a trace is started before the profiler flushes out and finishes after.

That also means that if a trace is more than one-minute long, it's certain that
the first part of its profiling samples will have the initial ("wrong")
resource being set, unfortunately.

### CODE REVIEW

**Review by brettlangdon — APPROVED** [review]
> That also means that if a trace is more than one-minute long, it's certain that
the first part of its profiling samples will have the initial ("wrong")
resource being set, unfortunately.

it is less about the length of the trace and a race condition between trace start, profiling flushing, and trace end right?

**Review by nizox — APPROVED** [review]

File: ddtrace/profiling/event.py [hunk 1]
```diff
@@ -44,7 +45,28 @@ class StackBasedEvent(SampleEvent):
...
+                self.trace_type = span._local_root.span_type
+            span._on_finish_callbacks.append(self._update_trace_resource)
+
+    def _update_trace_resource(
+        self,
+        span,  # type: ddspan.Span
+    ):
+        # type: (...) -> None
+        if span._local_root is not None:
+            self.trace_resource = span._local_root.resource
```
Comment (nizox) [inline ddtrace/profiling/event.py:72]: What about the span type ? I don't think the current integrations set it after a span creation though.

Comment (jd) [reply ddtrace/profiling/event.py:72]: We could probably do this here too. My point was to do the bare minimum, and I don't think as you're saying, that the type is set after. :)

**Comment by jd** [comment]
> it is less about the length of the trace and a race condition between trace start, profiling flushing, and trace end right?

Yes, you're just sure to hit the race condition at least once if your trace ie > 60s.

**Pull request merged**